### ISSUE TITLE
istioctl proxy-config command hangs in 1.2.0

### ISSUE THREAD

**Issue by Stono** [istio/istio#15126]
**Bug description**
istioctl proxy-config <anything> hangs indefinitely on 1.2.0, eg: `istioctl proxy-config endpoint platform-testing-6c69d48bb9-c7ht2.platform-testing`

**Affected product area (please put an X in all that apply)**

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[x] Test and Release
[x] User Experience
[x] Developer Infrastructure

**Expected behavior**
istioctl proxy-config to work

**Steps to reproduce the bug**
Use 1.2.0 and it won't work, tested on 3 different clusters.

**Version (include the output of `istioctl version --remote` and `kubectl version`)**
Istio 1.2.0, GKE 1.13.6

**How was Istio installed?**
Helm

**Environment where bug was observed (cloud vendor, OS, etc)**
GKE

**Comment by esnible** [comment]
The command isn't hanging for me.  I am also using 1.2.0.

Until I can reproduce it I will need your help in understanding what is wrong.  The command does the equivalent of

```
kubectl -n <ns> exec <pod> -c istio-proxy -- /usr/local/bin/pilot-agent request GET config_dump
```

What output do you get from this command?

If that command hangs, tell me what you get from `kubectl -n <ns> exec <pod> -c istio-proxy -- ps -ef` and `kubectl -n <ns> get pod <pod>`

**Comment by Stono** [comment]
@esnible thanks for your reply.

> kubectl -n <ns> exec <pod> -c istio-proxy -- /usr/local/bin/pilot-agent request GET config_dump

This command completes, but **does not exit**, eg I see the stdout config dump, but it's stuck until I CTRL+C it.

<img width="357" alt="Screenshot 2019-06-25 at 14 43 58" src="https://user-images.githubusercontent.com/1486729/60103535-afddac00-9757-11e9-8cbc-8e2fac639c14.png">

I've tested this with `istioctl` version `1.2.0` against version `1.1.9` and `1.2.0` pilots.  So it does seem to be an issue with the istioctl `1.2.0` binary

**Comment by esnible** [comment]
I am at a bit of a loss because I haven't seen `kubectl exec` pause like that.  I need to understand if the problem is with Istio's _/usr/local/bin/pilot-agent_ or with configuration specific to you.

Do `kubectl -n <ns> exec -it <pod> -c istio-proxy bash` to get a $ prompt, then run `/usr/local/bin/pilot-agent request GET config_dump` from that prompt and let me know if it completes.

Your comment says the problem is the _istioctl_ 1.2.0 binary but have you tried this with _istioctl_ 1.1.x?  (For testing I keep an Istio 1.1.x binary around, renamed _istioctl11_.  Just download an Istio 1.1.x and use _bin/istioctl_ from it).  I know in 1.2 we made changes on `istioctl proxy-status` but I believe the code for `istioctl proxy-config` is unchanged.

**Comment by Stono** [comment]
thanks again for the replies, i've actually left the office for the day so will continue debugging either later tonight or early tomorrow and get back to you with more info!

**Comment by GregHanson** [comment]
I tried the 1.2.0 istioctl binary on my Mac for a combination of gateways and apps and not seeing the command hang at all.  I think @esnible and I are both running against and IBM cloud deployment - maybe this is specific to GKE?  @Stono, hate to ask, but did you do a sanity check and verify that you are using the 1.2.0 istioctl binary and that you don't have an old sym link or `PATH` variable that points or an older istioctl binary?

**Comment by Stono** [comment]
We do quite a few custom things :-) it could well be user error. @esnible questions are good - will continue debugging shortly and let you know!

**Comment by Stono** [comment]
@GregHanson @esnible - thanks both for your input, I've resolved the issue.  When @esnible told me that it effectively runs `pilot-agent` within the container it reminded me of a script that we have.

Just so you know... there are a couple of issues with `pilot-agent`.  The first is that it doesn't play with cronjobs, and calling the envoy `quitquitquit` endpoint [leaves pilot-agent in a broken state](https://github.com/istio/istio/issues/15041).  The second is we want pilot-agent to only terminate [when the main application has stopped listening](https://github.com/istio/istio/issues/12183).

As a result, we have a script named `pilot-agent-agent` (copyright @howardjohn :D): https://gist.github.com/Stono/7bf43a94ee9a132ecbea869f60ebb28d which takes of this for us.

In our `1.2.0` (not `1.1.9`) branch, this script overwrote the `pilot-agent` binary in the container, and moved the original to `pilot-agent-agent`.  Now that's the problem, `istioctl proxy-status` was exec'ing and calling `pilot-agent`, which was actually running that script, which was actually then waiting for the application to shut down (which it obviously never would have done).

I've since updated this to simply be a `pilot-agent-agent` script, and updated the `ENTRYPOINT` on our custom `istio-proxy` image, and I am once again able to run `istioctl` commands.

In summary; user-error, but I just wanted you to understand the reasoning why :-) 

There are a few PR's in place which should hopefully remove my requirement for this in the near future anyway! 

Thank you both.
